discovery-utils/src/discovery/mod.rs

```diff
@@ -0,0 +1,252 @@
+        let discovery_handler_temp_dir = Builder::new()
+            .prefix("discovery-handlers")
+            .tempdir()
+            .unwrap();
+        let discovery_handler_temp_dir_path = discovery_handler_temp_dir.path().join(socket_name);
+        (
+            discovery_handler_temp_dir
+                .path()
+                .to_str()
+                .unwrap()
+                .to_string(),
+            discovery_handler_temp_dir_path
+                .to_str()
+                .unwrap()
+                .to_string(),
+        )
+    }
+
+    pub async fn run_mock_discovery_handler(
+        discovery_handler_dir: &str,
+        discovery_handler_endpoint: &str,
+    ) -> tokio::task::JoinHandle<()> {
+        let discovery_handler = MockDiscoveryHandler {};
+        let discovery_handler_dir_string = discovery_handler_dir.to_string();
+        let discovery_handler_endpoint_string = discovery_handler_endpoint.to_string();
+        let handle = tokio::spawn(async move {
+            super::server::internal_run_discovery_server(
+                discovery_handler,
+                &discovery_handler_endpoint_string,
+                &discovery_handler_dir_string,
+            )
+            .await
+            .unwrap();
+        });
+
+        // Try to connect in loop until first thread has served Discovery Handler
+        unix_stream::try_connect(discovery_handler_endpoint)
+            .await
+            .unwrap();
+        handle
+    }
+}
+
+pub mod server {
+    use super::v0::discovery_server::{Discovery, DiscoveryServer};
+    use akri_shared::uds::unix_stream;
+    use futures::stream::TryStreamExt;
+    use log::info;
+    use std::path::Path;
+    use tokio::net::UnixListener;
+    use tonic::transport::Server;
+
+    pub async fn run_discovery_server(
+        discovery_handler: impl Discovery,
+        discovery_endpoint: &str,
+    ) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+        internal_run_discovery_server(
+            discovery_handler,
+            discovery_endpoint,
+            super::DISCOVERY_HANDLER_PATH,
+        )
+        .await
+    }
+
+    /// Creates a DiscoveryServer for the given Discovery Handler at the specified endpoint
+    /// Verifies the endpoint by checking that it is in the discovery handler directory if it is
+    /// UDS or that it is a valid IP address and port.
+    pub async fn internal_run_discovery_server(
+        discovery_handler: impl Discovery,
+        discovery_endpoint: &str,
+        discovery_handler_directory: &str,
+    ) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+        info!("internal_run_discovery_server - entered");
+
+        if discovery_endpoint.starts_with(discovery_handler_directory) {
+            tokio::fs::create_dir_all(Path::new(&discovery_endpoint[..]).parent().unwrap()).await?;
+            // Delete socket if it already exists
+            std::fs::remove_file(discovery_endpoint).unwrap_or(());
```

Comment:
maybe this is where we can catch multiple agents running on a single node?  if we can't delete the file because another agent is using it, maybe we bubble the error up and kill this agent?

Reply:
from another comment, it seems like we already catch multiple agents running on a single node when the socket is listened on.